### PR TITLE
NODE-2692: Increase linting strictness

### PR DESCRIPTION
**Pull request by nbbeeken**
Remove most of the disabled lint rules and correct the resulting errors.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -92,7 +92,7 @@
     "build:docs": "npm run build:dts && typedoc",
     "check:bench": "node test/benchmarks/driverBench",
     "check:coverage": "nyc npm run check:test",
-    "check:lint": "eslint -v && eslint --max-warnings=0 --ext '.js,.ts' src test",
+    "check:lint": "tsc -v && tsc --noEmit && eslint -v && eslint --max-warnings=0 --ext '.js,.ts' src test",
```
Comment (nbbeeken) [inline package.json:95]: It seems apt that our compilation should be checked as part of linting

File: src/collection.ts [hunk 2]
```diff
@@ -126,7 +126,7 @@ export interface CollectionOptions
 
 /** @internal */
 export interface CollectionPrivate {
-  pkFactory: PkFactory | typeof ObjectId;
+  pkFactory: PkFactory;
```
Comment (nbbeeken) [inline src/collection.ts:129]: Our ObjectId class does have a `createPk` method it's just missing from the community types so I prefer here to cast in the constructor which we can revisit when we have correct typing.

File: src/collection.ts [hunk 3]
```diff
@@ -1652,16 +1649,30 @@ export class Collection implements OperationParent {
...
+  findAndRemove(query: Document): Promise<Document>;
+  findAndRemove(query: Document, sort: Sort): Promise<Document>;
+  findAndRemove(query: Document, sort: Sort, options: FindAndModifyOptions): Promise<Document>;
+  findAndRemove(
+    query: Document,
+    ...args: any[]
+  ): // TODO: Use labeled tuples when api-extractor supports TS 4.0
+  /*sort?: Sort | FindAndModifyOptions | Callback<Document>, */
+  /*options?: FindAndModifyOptions | Callback<Document>, */
+  /*callback?: Callback<Document>*/
```
Comment (nbbeeken) [inline src/collection.ts:1670]: At first glance this might look like comment garbage, and I can get rid of it if we don't want this to potentially sit here and get stale.
I would like to type args with
```typescript
...args: [sort?: Sort | FindAndModifyOptions | Callback<Document>, options?: FindAndModifyOptions | Callback<Document>, callback?: Callback<Document>]
```
Making use of the new labeled tuples gives a friendly auto complete and allows "optional" tuple members with the `?:` symbol, unfortunately you can't replicate that with a plain tuple, and currently api-extractor doesn't have the logic to parse this new TS feature.

File: src/collection.ts [hunk 4]
```diff
@@ -1687,26 +1698,23 @@ export class Collection implements OperationParent {
...
-    const args = Array.prototype.slice.call(arguments, 3);
-    callback = typeof args[args.length - 1] === 'function' ? args.pop() : undefined;
-    reduce = args.length ? args.shift() : null;
-    finalize = args.length ? args.shift() : null;
-    command = args.length ? args.shift() : null;
-    options = args.length ? args.shift() || {} : {};
+    condition: Document,
+    initial: Document,
+    // TODO: Use labeled tuples when api-extractor supports TS 4.0
+    ...args: [/*reduce?:*/ any, /*finalize?:*/ any, /*command?:*/ any, /*callback?:*/ Callback]
```
Comment (nbbeeken) [inline src/collection.ts:1706]: I can leave the tuple here since most of the arguments are `any` its too permissive to error.

File: src/mongo_client.ts [hunk 5]
```diff
@@ -427,13 +419,9 @@ export class MongoClient extends EventEmitter implements OperationParent {
...
   static connect(url: string, options: MongoClientOptions): Promise<MongoClient>;
   static connect(url: string, options: MongoClientOptions, callback: Callback<MongoClient>): void;
-  static connect(
-    url: string,
-    options?: MongoClientOptions | Callback<MongoClient>,
-    callback?: Callback<MongoClient>
-  ): Promise<MongoClient> | void {
-    if (typeof options === 'function') (callback = options), (options = {});
-    options = options || {};
+  static connect(url: string, ...args: any[]): Promise<MongoClient> | void {
```
Comment (nbbeeken) [inline src/mongo_client.ts:422]: This is another place for the labeled tuple style that we could use when api-extractor supports TS 4

File: src/operations/bulk_write.ts [hunk 6]
```diff
@@ -45,7 +45,7 @@ export class BulkWriteOperation extends OperationBase<BulkWriteOptions, BulkWrit
         : coll.initializeUnorderedBulkOp(options);
 
     // Do we have a collation
-    let collation = false;
+    const collation = false;
```
Comment (nbbeeken) [inline src/operations/bulk_write.ts:48]: collation is never changed, there's an if statement about 20 lines down that will never run, just wanted to get a second pair of eyes that this is correct and its okay to delete?

File: src/operations/common_functions.ts [hunk 7]
```diff
@@ -81,19 +82,23 @@ export function indexInformation(
...
 
   // no need to modify the docs if server sets the ObjectId
   if (forceServerObjectId === true) {
     return docs;
   }
 
-  return docs.map((doc: any) => {
-    if (forceServerObjectId !== true && doc._id == null) {
+  return docs.map(doc => {
+    if (doc._id == null) {
```
Comment (nbbeeken) [inline src/operations/common_functions.ts:101]: `forceServerObjectId` is either false or undefined after the check above asserts it can't be `true` so this was redundant

File: tsconfig.json [hunk 8]
```diff
@@ -25,6 +25,7 @@
     "inlineSources": true,
     // Prevents web types from being suggested by vscode.
     "types": ["node"],
+    "forceConsistentCasingInFileNames": true
```
Comment (nbbeeken) [inline tsconfig.json:28]: I don't expect us to ever make a file with capital letters in it but this is some free linting to prevent potential issues on windows

File: src/promise_provider.ts [hunk 9]
```diff
@@ -14,6 +14,7 @@ const store: PromiseStore = {
  */
 export class PromiseProvider {
   /** Validates the passed in promise library */
+  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
```
Comment (mbroadst) [inline src/promise_provider.ts:17]: why is this needed? I just checked in vscode and there were no errors in this file

File: src/read_preference.ts [hunk 10]
```diff
@@ -31,7 +33,9 @@ export interface ReadPreferenceOptions {
 }
 
 /** @public */
-export interface ReadPreferenceLikeOptions {
+export interface ReadPreferenceLikeOptions extends ReadPreferenceOptions {
+  readPreferenceTags?: TagSet[];
+  session?: ClientSession;
```
Comment (mbroadst) [inline src/read_preference.ts:38]: this looks wrong, session should have nothing to do with a `ReadPreferenceLikeOptions`

File: src/apm.ts [hunk 11]
```diff
@@ -42,12 +43,14 @@ export class Instrumentation extends EventEmitter {
...
+    } as MongoClient['connect'];
 
     if (typeof callback === 'function') callback(undefined, this);
   }
 
-  uninstrument() {
-    this.$MongoClient.prototype.connect = this.$prototypeConnect;
+  uninstrument(): void {
+    if (this.$MongoClient) {
+      this.$MongoClient.prototype.connect = this.$prototypeConnect as any;
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/apm.ts]: nope 🙂  Gone Away.

File: src/change_stream.ts [hunk 12]
```diff
@@ -209,6 +209,7 @@ export class ChangeStream extends EventEmitter {
       this.topology = parent.s.topology;
     } else if (parent instanceof MongoClient) {
       this.type = CHANGE_DOMAIN_TYPES.CLUSTER;
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/change_stream.ts:221]: I don't see an epic nor a ticket for this, I can make one.

File: src/collection.ts (hunk 2 above)
Comment (nbbeeken) [reply src/collection.ts:129]: Is this because ObjectID shouldn't have createPk?
cus wouldn't this just be:
```
class DefaultPkFactory {
  static createPk() { return ObjectId.createPk() }
}
```
Why the pass through class?

File: src/collection.ts (hunk 3 above)
Comment (nbbeeken) [reply src/collection.ts]: You do still make use of arguments in distinct to check the number of arguments provided. Doing a spread just makes this array explicit rather than the implicit global. I do agree though there isn't a benefit larger than that for now while we can't do the labeled tuples but they do provide nice hinting so something to consider for the future.

File: src/error.ts [hunk 13]
```diff
@@ -177,8 +177,9 @@ export class MongoParseError extends MongoError {
  */
 export class MongoTimeoutError extends MongoError {
   /** An optional reason context for the timeout, generally an error saved during flow of monitoring and selecting servers */
-  reason?: string;
+  reason?: unknown;
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/error.ts]: It appears to be only `TopologyDescription` good catch!

File: src/mongo_client.ts (hunk 5 above)
Comment (nbbeeken) [reply src/mongo_client.ts]: Reverted this, it was left over from an attempt to use the labeled tuples.

File: src/operations/group.ts [hunk 14]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/operations/group.ts:1]: There's a lot of `any` left in here, and its kinda ugly to type without _bsontype existing on Code. I figured since this is deprecated API it was something we would be okay with leave untyped especially since its internal at this layer anyway.
This might actually go away actually with the annotations work

File: src/promise_provider.ts (hunk 9 above)
Comment (nbbeeken) [reply src/promise_provider.ts]: I was able to get rid of this but you should see an error for any function that uses `any` that's exported from a module.

File: src/read_preference.ts (hunk 10 above)
Comment (nbbeeken) [reply src/read_preference.ts]: Indeed I added a new type `ReadPreferenceFromOptions` that represents this better, its public for now but I believe it will be made internal in the annotations work.

File: src/gridfs-stream/upload.ts [hunk 15]
```diff
@@ -208,8 +209,9 @@ export class GridFSBucketWriteStream extends Writable {
     this.state.streamEnd = true;
 
     if (callback) {
+      const cb = callback;
       this.once(GridFSBucketWriteStream.FINISH, (result: GridFSFile) => {
-        callback!(undefined, result);
+        cb(undefined, result);
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/gridfs-stream/upload.ts]: There is a good reason:
```typescript
let callbackish: Function | undefined = () => console.log('yay')
if(callbackish){
  doLater(() => {
    callbackish() // error potentially nullable
  });
}
callbackish = undefined;
```
Since doLater runs async, its possible callback could be made null again by the time it runs.
Updated the code with your suggestion

File: src/connection_string.ts [hunk 16]
```diff
@@ -36,7 +37,7 @@ function matchesParentDomain(srvAddress: string, parentDomain: string): boolean
  * @param options - Optional user provided connection string options
  */
 function parseSrvConnectionString(uri: string, options: any, callback: Callback) {
-  const result: any = url.parse(uri, true);
+  const result: Document = url.parse(uri, true);
```
Comment (mbroadst) [inline src/connection_string.ts:40]:
```suggestion
  const result: ArbitraryOptions = url.parse(uri, true);
```

Comment (mbroadst) [reply src/connection_string.ts:40]: also bring on the :bike: :house: but should it really be called `AnyOptions`?

File: src/mongo_client.ts [hunk 17]
```diff
@@ -316,7 +308,7 @@ export class MongoClient extends EventEmitter implements OperationParent {
    * @see docs.mongodb.org/manual/reference/connection-string/
    */
   connect(): Promise<MongoClient>;
-  connect(callback: Callback<MongoClient>): void;
+  connect(callback: Callback<MongoClient> | undefined): void;
```
Comment (mbroadst) [inline src/mongo_client.ts:311]:
```suggestion
  connect(callback?: Callback<MongoClient>): void;
```

File: src/collection.ts [hunk 18]
```diff
@@ -213,23 +215,21 @@ export class Collection implements OperationParent {
 
   /**
    * The name of the database this collection belongs to
-   * @readonly
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/collection.ts:216]: If there's only a getter, its not just TS readonly but also runtime readonly too and TS knows to pick that up. The TS keyword can be used to create compiler errors if you want something to behave as readonly. And then there's the TSDoc `@readonly` which if you have a setter (maybe for internal use) but you want the documentation to reflect it as readonly you can use this annotation to do so.

File: src/connection_string.ts (hunk 16 above)
Comment (nbbeeken) [reply src/connection_string.ts]: I'm fine with that, you could say the name is... _arbitrary_ 🥁

File: src/gridfs-stream/upload.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (mbroadst) [reply src/gridfs-stream/upload.ts]: right, but it's unlikely that it is reassigned. It would be nice if there was some sort of way to hint to the compiler that if a value was provided for `callback` then it would be immutable for the duration of the methods scope. :shrug:

File: src/read_preference.ts [hunk 19]
```diff
@@ -31,17 +33,25 @@ export interface ReadPreferenceOptions {
...
+        mode?: ReadPreferenceMode;
+        preference?: ReadPreferenceMode;
         tags: TagSet[];
         maxStalenessSeconds: number;
       };
 }
 
+/** @public */
+export interface ReadPreferenceFromOptions {
+  session?: ClientSession;
```
Comment (mbroadst) [inline src/read_preference.ts:49]: I'm still confused as to why `session` makes the cut here? It doesn't seem relevant to a read preference

Comment (nbbeeken) [reply src/read_preference.ts:49]: Maybe I should make another type, but it comes from the `resolve` method that uses the session to determine the read pref from a transaction if there is one.

File: src/operations/common_functions.ts [hunk 20]
```diff
@@ -81,28 +82,32 @@ export function indexInformation(
...
   if (forceServerObjectId === true) {
     return docs;
   }
 
-  return docs.map((doc: any) => {
-    if (forceServerObjectId !== true && doc._id == null) {
-      doc._id = coll.s.pkFactory.createPk();
+  return docs.map(doc => {
+    if (doc._id == null) {
+      doc._id = coll.s.pkFactory();
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/operations/common_functions.ts]: Reverted 👍

File: src/read_preference.ts (hunk 19 above)
Comment (mbroadst) [reply src/read_preference.ts:49]: ah! okay thanks, I forgot about that.
